### PR TITLE
♻️ Fixes thomas

### PR DESCRIPTION
**Pull request by Thomas-Smets**

### CODE REVIEW

**Comment by Thomas-Smets** [comment]
Test will only work after https://github.com/arcadia-finance/accounts-v2/actions/runs/6888429622/job/18737411587?pr=56 is merged

File: src/interfaces/IAccount.sol [hunk 1]
```diff
@@ -43,7 +43,7 @@ interface IAccount {
         external
         returns (address, uint256);
 
-    function startLiquidation()
+    function startLiquidation(address liquidationInitiator)
```
Comment (j-vp) [inline src/interfaces/IAccount.sol:46]: natspec

File: src/interfaces/ILendingPool.sol [hunk 2]
```diff
@@ -51,19 +51,13 @@ interface ILendingPool {
...
     function repay(uint256 amount, address account) external;
-    function auctionRepay(
-        uint256 startDebt,
-        address initiator,
-        address originalOwner,
-        uint256 amount,
-        address account,
-        address bidder
-    ) external returns (bool);
+    function auctionRepay(uint256 startDebt, address originalOwner, uint256 amount, address account, address bidder)
```
Comment (j-vp) [inline src/interfaces/ILendingPool.sol:54]: natspec of functions

File: src/Creditor.sol [hunk 3]
```diff
@@ -90,5 +90,5 @@ abstract contract Creditor {
      * @notice Starts the liquidation of an account and returns the open position of the Account.
      * @return openPosition the open position of the Account
```
Comment (j-vp) [inline src/Creditor.sol:91]: `* @param liquidationInitiator The address that starts the liquidation flow.`

Comment (j-vp) [reply src/Creditor.sol:91]: + in LendingPool.sol the variable is named "initiator", in accounts also liquidationInitiator

File: src/LendingPool.sol [hunk 4]
```diff
@@ -523,68 +521,59 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
-     * @param from The address of the caller performing the repayment.
-     * @dev This function transfers tokens from the `from` address to the contract and updates the user's account balance accordingly.
-     * @dev This internal function allows the caller to repay a specified amount of debt for a user.
+     * @notice Repays debt via an auction.
+     * @param startDebt The amount of debt of the Account the moment the liquidation was initiated.
+     * @param originalOwner The address of the Account owner.
+     * @param amount The amount of debt repaid by a bidder during the auction.
+     * @param account The contract address of the Arcadia Account backing the loan.
+     * @param bidder The address of the bidder.
+     * @return earlyTerminate Bool indicating of the full amount of debt was repaid.
```
Comment (j-vp) [inline src/LendingPool.sol:551]:
```suggestion
     * @return earlyTerminate Bool indicating whether the full amount of debt was repaid.
```

File: src/LendingPool.sol [hunk 5]
```diff
@@ -523,68 +521,59 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
+    {
         // Need to transfer before burning debt or ERC777s could reenter.
         // Address(this) is trusted -> no risk on re-entrancy attack after transfer.
-        asset.safeTransferFrom(from, address(this), amount);
+        asset.safeTransferFrom(bidder, address(this), amount);
 
-        _withdraw(repayAmount, account, account);
+        uint256 accountDebt = maxWithdraw(account);
+        if (accountDebt < amount) {
+            // The amount recovered by selling assets during the auction is bigger as the total debt of the Account.
```
Comment (j-vp) [inline src/LendingPool.sol:567]:
```suggestion
            // The amount recovered by selling assets during the auction is bigger than the total debt of the Account.
```

File: src/LendingPool.sol [hunk 6]
```diff
@@ -523,68 +521,59 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
+        returns (bool earlyTerminate)
+    {
         // Need to transfer before burning debt or ERC777s could reenter.
         // Address(this) is trusted -> no risk on re-entrancy attack after transfer.
-        asset.safeTransferFrom(from, address(this), amount);
+        asset.safeTransferFrom(bidder, address(this), amount);
 
-        _withdraw(repayAmount, account, account);
+        uint256 accountDebt = maxWithdraw(account);
+        if (accountDebt < amount) {
```
[the remark this answers is not in the document]
Comment (Thomas-Smets) [reply src/LendingPool.sol:566]: Yes, but is for a separate PR, it's related to what I discussed this morning with @acollette

File: src/LendingPool.sol [hunk 7]
```diff
@@ -905,61 +891,49 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
+            // Increase the realised liquidity for the original owner.
+            realisedLiquidityOf[originalOwner] += surplus;
 
-        // Calculate the bad debt.
-        if (surplus == 0) {
-            uint256 badDebt;
-            uint256 remainder;
+            totalRealisedLiquidity = SafeCastLib.safeCastTo128(uint256(totalRealisedLiquidity) + rewardsAndSurplus);
+        } else {
+            // openDebt equals startDebt + startDebtInterest + liquidationInitiatorReward + auctionTerminationReward + liquidationFee + interests - bids.
```
Comment (j-vp) [inline src/LendingPool.sol:911]: Difference between startDebtInterest and interest?

File: src/LendingPool.sol [hunk 8]
```diff
@@ -905,61 +891,49 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
+            uint256 openDebt = maxWithdraw(account);
             if (openDebt > auctionTerminationReward + liquidationFee) {
```
Comment (j-vp) [inline src/LendingPool.sol:913]: Does this need to include the initiatorReward? Or is that already settled?

File: src/LendingPool.sol [hunk 9]
```diff
@@ -905,61 +891,49 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
+                uint256 badDebt;
                 unchecked {
                     badDebt = openDebt - auctionTerminationReward - liquidationFee;
                 }
-                totalRealisedLiquidity =
-                    SafeCastLib.safeCastTo128(uint256(totalRealisedLiquidity) + liquidationInitiatorReward - badDebt);
+                totalRealisedLiquidity = SafeCastLib.safeCastTo128(uint256(totalRealisedLiquidity) - badDebt);
```
Comment (j-vp) [inline src/LendingPool.sol:918]:
```suggestion
                totalRealisedLiquidity = totalRealisedLiquidity - badDebt;
```
right side can't be bigger than total totalRealisedLiquidity which is an 128

File: src/LendingPool.sol [hunk 10]
```diff
@@ -905,61 +891,49 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
                 _processDefault(badDebt);
             } else {
+                uint256 remainder;
                 if (openDebt >= liquidationFee) {
                     remainder = (liquidationFee + auctionTerminationReward) - openDebt;
```
Comment (j-vp) [inline src/LendingPool.sol:923]: Chances of underflow if openDebt >> auctionTerminationReward?

File: src/LendingPool.sol (hunk 8 above)
Comment (Thomas-Smets) [reply src/LendingPool.sol:913]: No this is when: openDebt > auctionTerminationReward + liquidationFee -> we will pay no auctionTerminationReward and liquidationFee

File: src/LendingPool.sol [hunk 11]
```diff
@@ -905,61 +891,49 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
                     remainder = (liquidationFee - openDebt) + auctionTerminationReward;
                     // Increase the realised liquidity for the terminator.
                     realisedLiquidityOf[terminator] += auctionTerminationReward;
-                    // Synchronize the liquidation fee with liquidity providers.
+                    // Distribute the liquidation fee with liquidity providers.
                     _syncLiquidationFeeToLiquidityProviders(remainder - auctionTerminationReward);
                 }
-                totalRealisedLiquidity =
-                    SafeCastLib.safeCastTo128(uint256(totalRealisedLiquidity) + liquidationInitiatorReward + remainder);
+                totalRealisedLiquidity = SafeCastLib.safeCastTo128(uint256(totalRealisedLiquidity) + remainder);
```
Comment (j-vp) [inline src/LendingPool.sol:932]:
```suggestion
                totalRealisedLiquidity = totalRealisedLiquidity + remainder;
```

File: src/LendingPool.sol [hunk 12]
```diff
@@ -1051,26 +1025,33 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
         whenLiquidationNotPaused
         processInterests
         returns (uint256 startDebt)
     {
-        //Only Accounts can have debt, and debtTokens are non-transferrable.
-        //Hence by checking that the balance of the address passed as Account is not 0, we know the address
-        //passed as Account is indeed a Account and has debt.
+        // Only Accounts can have debt, and debtTokens are non-transferrable.
+        // Hence by checking that the balance of the msg.sender is not 0,
+        // we know that the sender is indeed a Account and has debt.
```
Comment (j-vp) [inline src/LendingPool.sol:1037]:
```suggestion
        // we know that the sender is indeed an Account and has debt.
```

File: src/LendingPool.sol [hunk 13]
```diff
@@ -1051,26 +1025,33 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
         (uint256 liquidationInitiatorReward, uint256 closingReward, uint256 liquidationPenalty) =
             _calculateRewards(startDebt);
 
-        // Mint extra debt towards the Account (as incentives should be considered in order to bring Account to a healthy state)
+        // Mint the liquidation incentives as extra debt towards the Account.
         _deposit(liquidationInitiatorReward + liquidationPenalty + closingReward, msg.sender);
 
+        // Increase the realised liquidity for the initiator.
+        realisedLiquidityOf[initiator] += liquidationInitiatorReward;
+        totalRealisedLiquidity = SafeCastLib.safeCastTo128(uint256(totalRealisedLiquidity) + liquidationInitiatorReward);
```
Comment (j-vp) [inline src/LendingPool.sol:1051]:
```suggestion
        totalRealisedLiquidity = totalRealisedLiquidity + liquidationInitiatorReward;
```

File: src/LendingPool.sol [hunk 14]
```diff
@@ -1051,26 +1025,33 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
      * @dev Only Accounts with non-zero balances can have debt, and debtTokens are non-transferrable.
```
Comment (j-vp) [inline src/LendingPool.sol:1025]: Split up the book above this line in multiple lines pls

File: src/LendingPool.sol [hunk 15]
```diff
@@ -1114,7 +1095,7 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
      * @notice Sets the liquidation weights.
      * @param initiatorRewardWeight_ Fee paid to the Liquidation Initiator.
      * @param penaltyWeight_ Penalty paid by the Account owner to the Creditor.
-     * @dev Each weight has 2 decimals precision (50 equals 0,5 or 50%).
+     * @dev Each weight has 4 decimals precision (50 equals 0,005 or 0,5%).
```
Comment (j-vp) [inline src/LendingPool.sol:1098]: Is this already changed?

File: src/LendingPool.sol (hunk 12 above)
Comment (Thomas-Smets) [reply src/LendingPool.sol:1037]: fixed that already in other branch

File: src/LendingPool.sol (hunk 15 above)
Comment (Thomas-Smets) [reply src/LendingPool.sol:1098]: Yes

File: src/LendingPool.sol [hunk 16]
```diff
@@ -523,68 +521,59 @@ contract LendingPool is LendingPoolGuardian, Creditor, DebtToken, InterestRateMo
...
+        uint256 accountDebt = maxWithdraw(account);
+        if (accountDebt < amount) {
+            // The amount recovered by selling assets during the auction is bigger as the total debt of the Account.
+            // -> Terminate the auction and make the surplus available to the Account-Owner.
+            earlyTerminate = true;
+            _settleLiquidation(account, originalOwner, startDebt, bidder, (amount - accountDebt));
+            amount = accountDebt;
+        }
+
+        _withdraw(amount, account, account);
```
Comment (j-vp) [inline src/LendingPool.sol:574]: if amount > accountDebt, will this not underflow?

File: src/Liquidator.sol [hunk 17]
```diff
@@ -307,11 +291,7 @@ contract Liquidator is Owned, ILiquidator {
         }
     }
 
-    function bid(address account, uint256[] memory assetAmounts, uint256[] memory assetIds, bool endAuction)
-        external
-        payable
-        nonReentrant
-    {
+    function bid(address account, uint256[] memory assetAmounts, uint256[] memory assetIds, bool endAuction) external {
```
Comment (j-vp) [inline src/Liquidator.sol:294]: natspec

File: src/Liquidator.sol [hunk 18]
```diff
@@ -474,19 +445,19 @@ contract Liquidator is Owned, ILiquidator {
     }
 
     function _knockDown(address account, AuctionInformation storage auctionInformation_) internal {
+        // Set the inAuction flag to false.
+        auctionInformation[account].inAuction = false;
+
         (bool success,,) = IAccount(account).isAccountHealthy(0, 0);
```
Comment (j-vp) [inline src/Liquidator.sol:451]: do we still need this with (0,0)?

File: src/Liquidator.sol [hunk 19]
```diff
@@ -474,19 +445,19 @@ contract Liquidator is Owned, ILiquidator {
...
         if (!success) revert Liquidator_AccountNotHealthy();
 
         uint256 startDebt = uint256(auctionInformation_.startDebt);
```
Comment (j-vp) [inline src/Liquidator.sol:454]:
```suggestion
        uint256 startDebt = auctionInformation_.startDebt;
```

**Review by j-vp — APPROVED** [review]
lgtm, open comments can be solved in other pr

File: src/Liquidator.sol (hunk 18 above)
Comment (Thomas-Smets) [reply src/Liquidator.sol:451]: yes different PR

Comment (Thomas-Smets) [reply src/Liquidator.sol:451]: https://arcadiafinance.atlassian.net/browse/AF-376

File: src/LendingPool.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (Thomas-Smets) [reply src/LendingPool.sol:566]: https://arcadiafinance.atlassian.net/browse/AF-377

**Pull request merged**